### ISSUE TITLE
Champion "permit pattern match of expression of type T:base with type derived"

### ISSUE THREAD

**Issue by gafter** [dotnet/csharplang#154]
- [ ] Proposal added
- [ ] Discussed in LDM
- [ ] Decision in LDM
- [ ] Finalized (done, rejected, inactive)
- [ ] Spec'ed

See https://github.com/dotnet/roslyn/issues/16195#issuecomment-270799191

**Comment by axel-habermaier** [comment]
This is a very surprising limitation. It severely limits the usefulness of pattern matching in generic code, making many of the changes I was hoping to make to my code base impossible.

**Comment by gafter** [comment]
@axel-habermaier cast the switch expression to `object` to work around this issue.

**Comment by alrz** [comment]
I think this is also related here (https://github.com/dotnet/roslyn/issues/17746)
> ```cs
> void M<T>(T o)
> {
>   if (o is X x)
>    ;
> }
> ```
Ideally, `e is T x` should be valid anywhere `e is T` is.

**Comment by aluanhaddad** [comment]
@gafter thanks for the work around but I really would like to have this work. I had the exact same  experience as @axel-habermaier and it was very frustrating especially as the error was the direct result of applying a built in Roslyn suggested refactoring inside the implementation of a pattern matching API I had written. I realize it's probably a bug in the analyzer but that really compounded the frustration since the `match` expression in VS 15 Preview had actually obviated the need for the API entirely.

**Comment by CyrusNajmabadi** [comment]
@aluanhaddad We discussed this in the LDM, and we believe this should be addressed.  @gafter's comment was more about what you could do in the meantime to unblock yourself.  We'll hopefully get this addressed in a upcoming release of VS (hopefully the next one :)).

**Comment by aluanhaddad** [comment]
@CyrusNajmabadi Thank you. I'm really glad to hear that this is likely to be resolved soon. That is excellent news!

**Comment by CyrusNajmabadi** [comment]
> I'm really glad to hear that this is likely to be resolved soon. 

Yup!  Though note the standard caveats about not being able to guarantee anything, and the future may change blah blah blah.  But i'm crossing my fingers that this smoothly slots into the 7.1 release.

**Comment by aluanhaddad** [comment]
> Though note the standard caveats about not being able to guarantee anything, and the future may change blah blah blah.

Indeed. Deadlines get missed. If that were a cardinal sin in our industry I probably wouldn't even have a job anymore.

**Comment by gafter** [comment]
I have added a proposal for this, and a PR exists for it in the Roslyn repo at https://github.com/dotnet/roslyn/pull/18784.

**Comment by shaggygi** [comment]
@gafter @jcouv was this implemented in 7.1 and now suppose to work in VS 2017?

**Comment by gafter** [comment]
Yes

**Comment by orthoxerox** [comment]
Shouldn't it be closed, then?

**Comment by jcouv** [comment]
There is a remaining workitem to update the language spec to reflect this feature. That's why the issue is kept open.

**Comment by shaggygi** [comment]
Thank you for reply

**Comment by reinseth** [comment]
The link to the proposal in the description is broken. Should be https://github.com/dotnet/csharplang/blob/master/proposals/csharp-7.1/generics-pattern-match.md

**Comment by jcouv** [comment]
@reinseth Fixed. Thanks
